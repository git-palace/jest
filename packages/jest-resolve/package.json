{
  "name": "jest-resolve",
  "version": "12.0.2",
  "repository": {
    "type": "git",
    "url": "https://github.com/facebook/jest.git"
  },
  "license": "BSD-3-Clause",
  "main": "src/index.js",
  "dependencies": {
    "jest-haste-map": "^12.0.2",
    "resolve": "^1.1.6"
  },
  "jest": {
    "testEnvironment": "node"
  },
  "scripts": {
    "test": "../../packages/jest-cli/bin/jest.js"
  }
}
