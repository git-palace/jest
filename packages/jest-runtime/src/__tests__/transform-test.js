/**
 * Copyright (c) 2014-present, Facebook, Inc. All rights reserved.
 *
 * This source code is licensed under the BSD-style license found in the
 * LICENSE file in the root directory of this source tree. An additional grant
 * of patent rights can be found in the PATENTS file in the same directory.
 *
 * @emails oncall+jsinfra
 */
'use strict';

jest
  .mock('graceful-fs')
  .mock('jest-resolve')
  .mock('jest-util')
  .mock('vm');

jest.mock(
  'test-preprocessor',
  () => {
    const stableStringify = require('json-stable-stringify');
    return {
      getCacheKey: jest.fn((content, filename, configStr) => 'ab'),
      process: (content, filename, config) => {
        return (
          `\nScript: ${content}\n` +
          `Path: ${filename}\n` +
          `Config: ${stableStringify(config, {space: 2})}`
        );
      },
    };
  },
  {virtual: true},
);

const getCachePath = (fs, config) => {
  for (const path in mockFs) {
    if (path.startsWith(config.cacheDirectory)) {
      return path;
    }
  }
  return null;
};

let config;
let fs;
let mockFs;
let object;
let transform;
let vm;

describe('transform', () => {

  const reset = () => {
    object = data => Object.assign(Object.create(null), data);

    vm = require('vm');

    mockFs = object({
      '/fruits/banana.js': [
        'module.exports = "banana";',
      ].join('\n'),
      '/fruits/kiwi.js': [
        'module.exports = () => "kiwi"',
      ].join('\n'),
      '/node_modules/react.js': [
        'module.exports = "react"',
      ].join('\n'),
    });

    fs = require('graceful-fs');
    fs.readFileSync = jest.fn((path, options) => {
      expect(options).toBe('utf8');

      if (mockFs[path]) {
        return mockFs[path];
      }

      throw new Error(`Cannot read path '${path}'.`);
    });
    fs.writeFileSync = jest.fn((path, data, options) => {
      expect(options).toBe('utf8');
      mockFs[path] = data;
    });

    fs.unlinkSync = jest.fn();
    fs.statSync = jest.fn(path => ({
      isFile: () => !!mockFs[path],
      mtime: {getTime: () => 42},
    }));

    require('jest-resolve').fileExists = jest.fn(path => !!mockFs[path]);

    config = {
      cache: true,
      cacheDirectory: '/cache/',
      name: 'test',
      preprocessorIgnorePatterns: ['/node_modules/'],
    };

    transform = require('../transform');
  };

  beforeEach(reset);

  it('transforms a file properly', () => {
    const response = transform('/fruits/banana.js', config);

    expect(response instanceof vm.Script);
    expect(vm.Script.mock.calls[0][0]).toMatchSnapshot();

    // no-cache case
    expect(fs.readFileSync.mock.calls.length).toBe(1);
    expect(fs.readFileSync).toBeCalledWith('/fruits/banana.js', 'utf8');

    // in-memory cache
    const response2 = transform('/fruits/banana.js', config);
    expect(response2).toBe(response);

    transform('/fruits/kiwi.js', config, {
      instrument: () => 'instrumented kiwi',
    });
    expect(vm.Script.mock.calls[1][0]).toMatchSnapshot();
  });

  it('uses the supplied preprocessor', () => {
    config = Object.assign(config, {
      scriptPreprocessor: 'test-preprocessor',
    });

    transform('/fruits/banana.js', config);

    expect(require('test-preprocessor').getCacheKey).toBeCalled();

    expect(vm.Script.mock.calls[0][0]).toMatchSnapshot();

    transform('/node_modules/react.js', config);
    expect(vm.Script.mock.calls[1][0]).toMatchSnapshot();
  });

  it('reads values from the cache', () => {
    const transformConfig = Object.assign(config, {
      scriptPreprocessor: 'test-preprocessor',
    });
    transform('/fruits/banana.js', transformConfig);

    const cachePath = getCachePath(mockFs, config);
    expect(fs.writeFileSync).toBeCalled();
    expect(fs.writeFileSync.mock.calls[0][0]).toBe(cachePath);

    // Cache the state in `mockFsCopy`
    const mockFsCopy = mockFs;
    jest.resetModuleRegistry();
    reset();

    // Restore the cached fs
    mockFs = mockFsCopy;
    transform('/fruits/banana.js', transformConfig);

    expect(fs.readFileSync.mock.calls.length).toBe(2);
    expect(fs.readFileSync).toBeCalledWith('/fruits/banana.js', 'utf8');
    expect(fs.readFileSync).toBeCalledWith(cachePath, 'utf8');
    expect(fs.writeFileSync).not.toBeCalled();

    // Don't read from the cache when `config.cache` is false.
    jest.resetModuleRegistry();
    reset();
    mockFs = mockFsCopy;
    transformConfig.cache = false;
    transform('/fruits/banana.js', transformConfig);

    expect(fs.readFileSync.mock.calls.length).toBe(1);
    expect(fs.readFileSync).toBeCalledWith('/fruits/banana.js', 'utf8');
    expect(fs.readFileSync).not.toBeCalledWith(cachePath, 'utf8');
    expect(fs.writeFileSync).toBeCalled();
  });
});
