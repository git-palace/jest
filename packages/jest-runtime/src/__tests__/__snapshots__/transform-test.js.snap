exports[`transform transforms a file properly 1`] = `
"({\"Object.<anonymous>\":function(module,exports,require,__dirname,__filename,global,jest,$JEST){module.exports = \"banana\";
}});"
`;

exports[`transform transforms a file properly 2`] = `
"({\"Object.<anonymous>\":function(module,exports,require,__dirname,__filename,global,jest,$JEST){instrumented kiwi
}});"
`;

exports[`transform uses the supplied preprocessor 1`] = `
"({\"Object.<anonymous>\":function(module,exports,require,__dirname,__filename,global,jest,$JEST){
Script: module.exports = \"banana\";
Path: /fruits/banana.js
Config: {
  \"cache\": true,
  \"cacheDirectory\": \"/cache/\",
  \"preprocessorIgnorePatterns\": [
    \"/node_modules/\"
  ],
  \"scriptPreprocessor\": \"test-preprocessor\"
}
}});"
`;

exports[`transform uses the supplied preprocessor 2`] = `
"({\"Object.<anonymous>\":function(module,exports,require,__dirname,__filename,global,jest,$JEST){module.exports = \"react\"
}});"
`;
